en: add doc for backup data to azblob (#1920)
pingcap/docs-tidb-operator: en/backup-to-azblob-using-br.md

---
title: Back up Data to Azure Blob Storage Using BR
summary: Learn how to back up data to Azure Blob Storage using BR.
---

# Back up Data to Azure Blob Storage Using BR

This document describes how to back up the data of a TiDB cluster in Kubernetes to Azure Blob Storage.

The backup method described in this document is implemented based on CustomResourceDefinition (CRD) in TiDB Operator. For the underlying implementation, [BR](https://docs.pingcap.com/tidb/stable/backup-and-restore-tool) is used to get the backup data of the TiDB cluster, and then send the data to Azure Blob Storage. BR stands for Backup & Restore, which is a command-line tool for distributed backup and recovery of the TiDB cluster data.

## Usage scenarios

If you have the following backup needs, you can use BR to make an [ad-hoc backup](#ad-hoc-backup) or [scheduled full backup](#scheduled-full-backup) of the TiDB cluster data to Azure Blob Storage.

> - To back up a large volume of data at a fast speed

Author: @xu21yingan PTAL. Does `1 TB` apply to BR in the Kubernetes environment?

> - To back up a large volume of data at a fast speed

Reviewer: Yes, but In keeping with the previous files, eg `[backup-to-aws-s3-using-br.md](https://github.com/pingcap/docs-tidb-operator/blob/master/en/backup-to-aws-s3-using-br.md)` `[backup-to-gcs-using-br.md](https://github.com/pingcap/docs-tidb-operator/blob/master/en/backup-to-gcs-using-br.md)`. Maybe all need to be adjusted.

> - To back up a large volume of data at a fast speed

Author: added in d4213ea

    {{< copyable "shell-regular" >}}

    ```shell
    kubectl apply -f backup-rbac.yaml -n test1
    ```

2. Grant permissions to the remote storage. You can grant permissions to Azure Blob Storage by two methods. For details, refer to [Azure account permissions](grant-permissions-to-remote-storage.md#azure-account-permissions).

3. For a TiDB version earlier than v4.0.8, you also need to complete the following preparation steps. For TiDB v4.0.8 or a later version, skip these preparation steps.

    1. Make sure that you have the `SELECT` and `UPDATE` privileges on the `mysql.tidb` table of the backup database so that the `Backup` CR can adjust the GC time before and after the backup.
    2. Create the `backup-demo1-tidb-secret` secret to store the account and password to access the TiDB cluster:

        {{< copyable "shell-regular" >}}

        ```shell
        kubectl create secret generic backup-demo1-tidb-secret --from-literal=password=${password} --namespace=test1
        ```

> ### Step 2: Perform an ad-hoc backup

Reviewer 2: ```suggestion
### Step 2. Perform an ad-hoc backup
```

```

</details>

<details>
<summary>Back up data of multiple tables using the table filter</summary>

The following example backs up data of the `db1.table1` table and `db1.table2` table.

```yaml
---
apiVersion: pingcap.com/v1alpha1
kind: Backup
metadata:
  name: demo1-backup-azblob
  namespace: test1
spec:
  backupType: full
  serviceAccount: tidb-backup-manager
  tableFilter:
  - "db1.table1"
  - "db1.table2"
  # ...
  br:
    cluster: demo1
    sendCredToTikv: false
    clusterNamespace: test1
  # Only needed for TiDB Operator < v1.1.10 or TiDB < v4.0.8
  # from:
    # host: ${tidb_host}
    # port: ${tidb_port}
    # user: ${tidb_user}
    # secretName: backup-demo1-tidb-secret
  azblob:
    secretName: azblob-secret-ad
    container: my-container
    bucket: my-bucket
    prefix: my-folder
```

</details>

## Scheduled full backup

You can set a backup policy to perform scheduled backups of the TiDB cluster, and set a backup retention policy to avoid excessive backup items. A scheduled full backup is described by a custom `BackupSchedule` CR object. A full backup is triggered at each backup time point. Its underlying implementation is the ad-hoc full backup.

> ### Step 1: Prepare for a scheduled full backup

Reviewer 2: ```suggestion
### Step 1. Prepare for a scheduled full backup
```

> The steps to prepare for a scheduled full backup are the same as that of [Prepare for an ad-hoc backup](#step-1-prepare-for-an-ad-hoc-backup).

Reviewer 2: ```suggestion
The steps to prepare for a scheduled full backup are the same as those of [Prepare for an ad-hoc backup](#step-1-prepare-for-an-ad-hoc-backup).
```

> ### Step 2: Perform a scheduled full backup

Reviewer 2: ```suggestion
### Step 2. Perform a scheduled full backup
```

    {{< copyable "shell-regular" >}}

    ```shell
    kubectl apply -f backup-scheduler-azblob.yaml
    ```

    The content of `backup-scheduler-azblob.yaml` is as follows:

    {{< copyable "" >}}

    ```yaml
    ---
    apiVersion: pingcap.com/v1alpha1
    kind: BackupSchedule
    metadata:
      name: demo1-backup-schedule-azblob
      namespace: test1
    spec:
      #maxBackups: 5
      #pause: true
      maxReservedTime: "3h"
      schedule: "*/2 * * * *"
      backupTemplate:
        backupType: full
        br:
          cluster: demo1
          sendCredToTikv: false
          clusterNamespace: test1
          # logLevel: info
          # statusAddr: ${status_addr}
          # concurrency: 4
          # rateLimit: 0
          # timeAgo: ${time}
          # checksum: true
        # Only needed for TiDB Operator < v1.1.10 or TiDB < v4.0.8
        from:
          host: ${tidb_host}
          port: ${tidb_port}
          user: ${tidb_user}
          secretName: backup-demo1-tidb-secret
        azblob:
          secretName: azblob-secret-ad
          container: my-container
          prefix: my-folder
    ```

> From the above content in `backup-scheduler-azblob.yaml`, you can see that the `backupSchedule` configuration consists of two parts. One is the unique configuration of `backupSchedule`, and the other is `backupTemplate`.

Reviewer 2: ```suggestion
From the preceding content in `backup-scheduler-azblob.yaml`, you can see that the `backupSchedule` configuration consists of two parts. One is the unique configuration of `backupSchedule`, and the other is `backupTemplate`.
```

- For the unique configuration of `backupSchedule`, refer to [BackupSchedule CR fields](backup-restore-cr.md#backupschedule-cr-fields).
- `backupTemplate` specifies the configuration related to the cluster and remote storage, which is the same as the `spec` configuration of [the `Backup` CR](backup-restore-cr.md#backup-cr-fields).

> After creating the scheduled full backup, use the following command to check the backup status:

Reviewer 2: ```suggestion
After creating the scheduled full backup, you can run the following command to check the backup status:
```

{{< copyable "shell-regular" >}}

```shell
kubectl get bks -n test1 -o wide
```

> You can use the following command to check all the backup items:

Reviewer 2: ```suggestion
You can run the following command to check all the backup items:
```

{{< copyable "shell-regular" >}}

```shell
kubectl get bk -l tidb.pingcap.com/backup-schedule=demo1-backup-schedule-azblob -n test1
```

## Delete the backup CR

> If you no longer need the backup CR, refer to [Delete the Backup CR](backup-restore-overview.md#delete-the-backup-cr).

Reviewer 2: ```suggestion
If you no longer need the backup CR, you can delete it by referring to [Delete the Backup CR](backup-restore-overview.md#delete-the-backup-cr).
```
